### PR TITLE
Adds cement and concrete, used to make concrete walls and floors.

### PR DESCRIPTION
**Pull request by tmtmtl30**
## About The Pull Request

Adds liquid cement and concrete, along with the stronger hexement and hexacrete. They can be used to make walls and floors.

Concrete walls and floors take time to harden, and must be destroyed instead of deconstructed. Concrete walls can only be removed by dealing enough damage to break them (mining tools deal more damage, sharp or hot items deal less), and concrete floors must be smashed using pickaxes or other mining tools. If you walk over a concrete floor while it's hardening, it'll leave a trail of footprints.

This PR includes minor changes to wall and reagent container code and a few other tweaks to ensure everything functions correctly. It isn't 100% finished, and there are a few things I need to test to ensure they work correctly.

Sprites are courtesy of @triplezeta :
image
Along the top, concrete walls and floors, the walls in various stages of dryness. Along the bottom, hexacrete walls and floors.

## Why It's Good For The Game

cement

## Changelog
:cl: tmtmtl30, tetrazeta
add: Pouring -- you can now pour out a portion of a beaker's contents by clicking on an object with a beaker while on help intent.
add: Liquid cement, made from heated carbon, hydrogen, oxygen, and water; it can be concentrated into hexement by adding phenol. Don't eat it.
add: Concrete, made by adding sand to liquid cement. Similarly, hexacrete can be made by adding sand to a beaker of hexement.
add: Concrete floors, made by pouring concrete or hexacrete onto catwalks over plating. Use a chisel to change how they look, but be careful not to step on them while they're wet.
add: Concrete walls, made by pouring concrete onto metal grilles. They take 30 seconds to fully harden, and will break if subjected to enough punishment.
add: Hexacrete walls, stronger concrete walls that take longer to harden, made by pouring hexacrete onto metal girders. Good for building explosion-resistant bunkers.
/:cl:

### CODE REVIEW

**Comment by triplezeta** [comment]
sliiight clarification, Azlan did the "smooth" concrete floor (far left). I did the rest of the sprites.

**Comment by Zevotech** [comment]
i do have some concerns about the beaker pouring thing, like how it interacts with tables and machines that you put beakers into.

otherwise yeah sure ill glaze a donut with cement

**Comment by tmtmtl30** [comment]
beakers are still placed on tables or into machines normally. that said, there are probably some weird edge cases i don't know about where pouring acts strange.

**Comment by Zevotech** [comment]
> beakers are still placed on tables or into machines normally. that said, there are probably some weird edge cases i don't know about where pouring acts strange.

Remember to consider things like mediborgs, service borgs and other non ordinary mobs that can access beakers.
also, harm intent will just splash a beaker onto something and while i understand that splashing cement onto a wall may not make some sense it wouldnt be too hard to just change the verb used.

**Comment by triplezeta** [comment]
the reason why splashing was intentionally not used was because splashing your *entire* beaker when you need to pour out an entire row of walls is... not optimal

**Comment by Zevotech** [comment]
hngh fuck good point. fucked up how that works. anyways yes the beaker storage module may not allow mediborgs to pour at all considering its only technically a beaker. either that or they can only pour with it.

**Comment by Sylphily** [comment]
Say, can you make it so that we can cement someone

**Comment by Apogee-dev** [comment]
i'd still love to have a special interaction for hitting someone with a spray or foam containing any of the cement or concrete reagents. GLOO gun type antics

**Comment by PrefabQuasar** [comment]
theres no method for rotating any of the asymetrical concrete floor tiles

**Comment by PrefabQuasar** [comment]
also radial menus for reskinning floor tiles feels extremely clunky, and makes building concrete, which is meant to be a faster way of building, significantly slower than using traditional tiles
also uses 4x as much metal, since you need 1 sheet of metal to make the rods for a single catwalk, and 1/4 a sheet of metal to make a tile

File: code/game/objects/structures/girders.dm [hunk 1]
```diff
@@ -332,6 +332,12 @@
 	girderpasschance = 0
 	max_integrity = 350
 
+/obj/structure/girder/reinforced/deconstruct(disassembled = TRUE)
+	if(!(flags_1 & NODECONSTRUCT_1))
+		var/remains = pick(/obj/item/stack/rods, /obj/item/stack/sheet/metal)
+		new remains(loc, 2)
```
Comment (Fikou) [inline code/game/objects/structures/girders.dm:338]: shouldnt this be only when not disassembled, and broken by force?

File: code/game/objects/items/tools/chisel.dm [hunk 2]
```diff
@@ -20,6 +20,9 @@
 	if(ismob(target))
 		return ..()
 	if(isturf(target))
+		if(istype(target, /turf/open/floor/concrete))
+			var/turf/open/floor/concrete/C = target
```
Comment (Fikou) [inline code/game/objects/items/tools/chisel.dm]: descript variable names please

File: code/game/turfs/closed/wall/misc_walls.dm [hunk 3]
```diff
@@ -30,13 +27,13 @@
...
 
-/turf/closed/wall/mineral/cult/artificer/break_wall()
-	new /obj/effect/temp_visual/cult/turf(get_turf(src))
-	return null //excuse me we want no runed metal here
-
-/turf/closed/wall/mineral/cult/artificer/devastate_wall()
+// no sheets, just a cult effect
+/turf/closed/wall/mineral/cult/artificer/create_sheets()
 	new /obj/effect/temp_visual/cult/turf(get_turf(src))
+	return null
```
Comment (Fikou) [inline code/game/turfs/closed/wall/misc_walls.dm]: any reason to not just use return?

File: code/game/turfs/closed/walls.dm [hunk 4]
```diff
@@ -43,6 +44,11 @@
 		fixed_underlay = string_assoc_list(fixed_underlay)
 		underlays += underlay_appearance
 
+/turf/closed/wall/update_overlays()
+	. = ..()
+	for(var/D in dent_decals)
```
Comment (Fikou) [inline code/game/turfs/closed/walls.dm]: descript variable name please

File: code/game/turfs/open/floor/conc_floor.dm [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+			return TRUE
+	return FALSE
+
+/turf/open/floor/concrete/proc/handle_shape(mob/user)
+	if(harden_lvl >= 0.8)
+		return FALSE
+	var/list/opts = list()
+
+	for(var/T in shape_types)
+		var/turf/open/floor/concrete/conc = T
```
Comment (Fikou) [inline code/game/turfs/open/floor/conc_floor.dm]: you can use typing with as anything, and descript variable name

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 6]
```diff
@@ -2453,3 +2453,104 @@
...
+	H.reagents.remove_reagent(src.type, amt)
+	H.reagents.add_reagent(concrete_type, amt)
+	return TRUE
+
+/datum/reagent/cement/expose_mob(mob/living/M, method=TOUCH, reac_volume, show_message = 1, touch_protection = 0)
+	. = ..()
+	if(!. || method != INGEST)
+		return
+	var/age = 6
+	if(ishuman(M))
```
Comment (Fikou) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm:2489]: and here

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 7]
```diff
@@ -2453,3 +2453,104 @@
...
+	if(istype(O, girder_type))
+		return concify_girder(O, volume)
+	if(istype(O, /obj/structure/catwalk))
+		return concify_catwalk(O, volume)
+
+/datum/reagent/concrete/proc/concify_girder(obj/O, volume)
+	if(volume < units_per_wall)
+		return
+	var/turf/T = get_turf(O)
+	var/turf/closed/wall/concrete/W = T.PlaceOnTop(wall_type)
```
Comment (Fikou) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm]: and here

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 8]
```diff
@@ -2453,3 +2453,104 @@
...
+	qdel(O)
+	return
+
+/datum/reagent/concrete/proc/concify_catwalk(obj/O, volume)
+	if(volume < units_per_floor)
+		return
+	var/turf/T = get_turf(O)
+	if(!istype(T, /turf/open/floor/plating))
+		return
+	var/turf/open/floor/concrete/F = T.PlaceOnTop(floor_type)
```
Comment (Fikou) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm]: and here

File: code/modules/reagents/reagent_containers.dm [hunk 9]
```diff
@@ -73,6 +73,16 @@
 	else
 		. += "<span class='notice'>The cap has been taken off. Alt-click to put a cap on.</span>"
 
+/obj/item/reagent_containers/is_injectable(mob/user, allowmobs = TRUE)
+	if(can_have_cap && cap_on)
+		return FALSE
+	. = ..()
```
Comment (Fikou) [inline code/modules/reagents/reagent_containers.dm]: use return ..() at the end of procs for readability, same with the lower one

File: code/modules/reagents/reagent_containers.dm [hunk 10]
```diff
@@ -100,6 +110,27 @@
...
+	if(!is_drainable() || !reagents.total_volume)
+		return
+	playsound(src, 'sound/items/glass_splash.ogg', 50, 1)
+	target.visible_message("<span class='notice'>[user] pours [src] onto [target].</span>")
+	log_combat(user, target, "poured [english_list(reagents.reagent_list)]", "in [AREACOORD(target)]")
+	log_game("[key_name(user)] poured [english_list(reagents.reagent_list)] on [target] in [AREACOORD(target)].")
+	var/frac = min(amount_per_transfer_from_this/reagents.total_volume, 1)
+	// don't use trans_to, because we're not ADDING it to the object, we're just... pouring it.
+	reagents.expose(target, TOUCH, frac)
+	for(var/reagent in reagents.reagent_list)
```
Comment (Fikou) [inline code/modules/reagents/reagent_containers.dm]: type it, use as anything

**Comment by Vexylius** [comment]
When filling a bucket from a sink, when it's full you start pouring water over the sink. Would say that's a bug

**Review by Fikou — CHANGES_REQUESTED** [review]

**Comment by Vexylius** [comment]
Please fix conflicts, everyone **NEEDS** cement

**Comment by Latentish** [comment]
I desperately request you finish this PR. I desire cement for my morning cereal.

**Comment by triplezeta** [comment]
pleading pleading pleading pleading 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺

File: code/game/turfs/closed/wall/misc_walls.dm (hunk 3 above)
Comment (tmtmtl30) [reply code/game/turfs/closed/wall/misc_walls.dm]: not particularly. will change

File: code/game/objects/structures/girders.dm (hunk 1 above)
Comment (tmtmtl30) [reply code/game/objects/structures/girders.dm:338]: this was to make rwalls drop slightly more metal when they get annihilated by explosions, i believe? you can see the default code for /obj/structure/girder/deconstruct() above -- note the similarity. girders don't use deconstruct() to handle deconstruction-via-tools, i think it's generally only called when their integrity falls to 0. not entirely sure what to do here

**Comment by tmtmtl30** [comment]
resolved merge conflicts, handled reviews

**Comment by Latentish** [comment]
Have any bugs been discovered from the testmerge? If not, I'd really love to see this merged for real so I can start using it in mapping.

**Comment by ZephyrTFA** [comment]
This has been test merged for about... *looks at watch* a month now. Is this good to merge?

**Comment by tmtmtl30** [comment]
> Is this good to merge?

probably? pouring is inherently kind of (very) janky but fixing that would likely involve a rework of the "reagent_containers" attack chain procs

**Review by triplezeta — APPROVED** [review]
as the person who made these sprites i approve:tm:

File: code/game/mecha/equipment/tools/mining_tools.dm [hunk 11]
```diff
@@ -64,13 +64,13 @@
 /turf/closed/wall/drill_act(obj/item/mecha_parts/mecha_equipment/drill/drill)
 	if(drill.do_after_mecha(src, 60 / drill.drill_level))
 		drill.log_message("Drilled through [src]", LOG_MECHA)
-		dismantle_wall(TRUE, FALSE)
+		dismantle_wall(TRUE)
```
Comment (Fikou) [inline code/game/mecha/equipment/tools/mining_tools.dm]: can you make this a named argument? just TRUE doesnt tell much

File: code/game/mecha/equipment/tools/mining_tools.dm [hunk 12]
```diff
@@ -64,13 +64,13 @@
...
 
 /turf/closed/wall/r_wall/drill_act(obj/item/mecha_parts/mecha_equipment/drill/drill)
 	if(drill.drill_level >= DRILL_HARDENED)
 		if(drill.do_after_mecha(src, 120 / drill.drill_level))
 			drill.log_message("Drilled through [src]", LOG_MECHA)
-			dismantle_wall(TRUE, FALSE)
+			dismantle_wall(TRUE)
```
Comment (Fikou) [inline code/game/mecha/equipment/tools/mining_tools.dm]: and here

File: code/modules/research/xenobiology/crossbreeding/burning.dm [hunk 13]
```diff
@@ -83,7 +83,7 @@ Burning extracts:
 
 /obj/item/slimecross/burning/metal/do_effect(mob/user)
 	for(var/turf/closed/wall/W in range(1,get_turf(user)))
-		W.dismantle_wall(1)
+		W.dismantle_wall(TRUE)
```
Comment (Fikou) [inline code/modules/research/xenobiology/crossbreeding/burning.dm]: and here

File: code/modules/projectiles/guns/misc/blastcannon.dm [hunk 14]
```diff
@@ -149,7 +149,7 @@
 			if(istype(loc, /turf/closed/wall))
 				var/turf/closed/wall/W = loc
 				if(prob(wallbreak_chance))
-					W.dismantle_wall(TRUE, TRUE)
+					W.dismantle_wall(TRUE)
```
Comment (Fikou) [inline code/modules/projectiles/guns/misc/blastcannon.dm]: and here

File: code/modules/projectiles/guns/misc/beam_rifle.dm [hunk 15]
```diff
@@ -453,7 +453,7 @@
 		if(prob(wall_devastate))
 			if(iswallturf(A))
 				var/turf/closed/wall/W = A
-				W.dismantle_wall(TRUE, TRUE)
+				W.dismantle_wall(TRUE)
```
Comment (Fikou) [inline code/modules/projectiles/guns/misc/beam_rifle.dm]: and here

File: code/modules/mob/living/simple_animal/hostile/space_dragon.dm [hunk 16]
```diff
@@ -116,7 +116,7 @@
 		if(do_after(src, timetotear, target = thewall))
 			if(istype(thewall, /turf/open))
 				return
-			thewall.dismantle_wall(1)
+			thewall.dismantle_wall(TRUE)
```
Comment (Fikou) [inline code/modules/mob/living/simple_animal/hostile/space_dragon.dm]: and here

File: code/game/turfs/closed/wall/mineral_walls.dm [hunk 17]
```diff
@@ -179,7 +179,7 @@
 		if(istype(W, /obj/item/hatchet) || istype(W, /obj/item/fireaxe))
 			duration /= 4 //Much better with hatchets and axes.
 		if(do_after(user, duration*10, target=src)) //Into deciseconds.
-			dismantle_wall(FALSE,FALSE)
+			dismantle_wall(FALSE)
```
Comment (Fikou) [inline code/game/turfs/closed/wall/mineral_walls.dm]: and here

File: code/game/turfs/closed/wall/reinf_walls.dm [hunk 18]
```diff
@@ -39,17 +39,13 @@
...
-	new /obj/item/stack/sheet/metal(src, 2)
-
 /turf/closed/wall/r_wall/attack_animal(mob/living/simple_animal/M)
 	M.changeNext_move(CLICK_CD_MELEE)
 	M.do_attack_animation(src)
 	if(!M.environment_smash)
 		return
 	if(M.environment_smash & ENVIRONMENT_SMASH_RWALLS)
-		dismantle_wall(1)
+		dismantle_wall(TRUE)
```
Comment (Fikou) [inline code/game/turfs/closed/wall/reinf_walls.dm]: and here

File: code/game/objects/items/pitchfork.dm [hunk 19]
```diff
@@ -100,6 +100,5 @@
 		var/turf/closed/wall/W = target
 		user.visible_message("<span class='danger'>[user] blasts \the [target] with \the [src]!</span>")
 		playsound(target, 'sound/magic/disintegrate.ogg', 100, TRUE)
-		W.break_wall()
-		W.ScrapeAway(flags = CHANGETURF_INHERIT_AIR)
+		W.dismantle_wall(FALSE)
```
Comment (Fikou) [inline code/game/objects/items/pitchfork.dm]: and here

File: code/game/turfs/closed/wall/conc_walls.dm [hunk 20]
```diff
@@ -0,0 +1,283 @@
...
+	update_icon()
+
+/turf/closed/wall/concrete/proc/alter_health(delta)
+	// 8x as vulnerable when unhardened
+	if(delta < 0)
+		delta *= 1 + 7*(1-harden_lvl)
+	health += delta
+	if(health <= 0)
+		// if damage put us 50 points or more below 0, we got proper demolished
+		dismantle_wall(health <= -50 ? TRUE : FALSE)
```
Comment (Fikou) [inline code/game/turfs/closed/wall/conc_walls.dm:130]: and here

File: code/game/turfs/closed/walls.dm [hunk 21]
```diff
@@ -53,47 +66,51 @@
...
-				dismantle_wall(0,1)
+				dismantle_wall(TRUE)
 			else
-				dismantle_wall(1,1)
-		if(3)
+				dismantle_wall(FALSE)
+		if(EXPLODE_LIGHT)
 			if (prob(hardness))
-				dismantle_wall(0,1)
+				dismantle_wall(FALSE)
```
Comment (Fikou) [inline code/game/turfs/closed/walls.dm]: and here

File: code/game/turfs/closed/walls.dm [hunk 22]
```diff
@@ -112,15 +129,14 @@
 			M.visible_message("<span class='danger'>[M.name] hits [src]!</span>", \
 							"<span class='danger'>You hit [src]!</span>", null, COMBAT_MESSAGE_RANGE)
 			if(prob(hardness + M.force) && M.force > 20)
-				dismantle_wall(1)
+				dismantle_wall(TRUE)
```
Comment (Fikou) [inline code/game/turfs/closed/walls.dm]: and here

File: code/game/turfs/closed/walls.dm [hunk 23]
```diff
@@ -132,20 +148,18 @@
 	M.do_attack_animation(src)
 	if((M.environment_smash & ENVIRONMENT_SMASH_WALLS) || (M.environment_smash & ENVIRONMENT_SMASH_RWALLS))
 		playsound(src, 'sound/effects/meteorimpact.ogg', 100, TRUE)
-		dismantle_wall(1)
+		dismantle_wall(TRUE)
```
Comment (Fikou) [inline code/game/turfs/closed/walls.dm]: and here

File: code/game/turfs/closed/walls.dm [hunk 24]
```diff
@@ -132,20 +148,18 @@
...
-	if(!arm)
-		return
-	if(arm.bodypart_disabled)
+	if(!arm || arm.bodypart_disabled)
 		return
 	if(prob(hardness))
 		playsound(src, 'sound/effects/meteorimpact.ogg', 100, TRUE)
 		user.say(pick(";RAAAAAAAARGH!", ";HNNNNNNNNNGGGGGGH!", ";GWAAAAAAAARRRHHH!", "NNNNNNNNGGGGGGGGHH!", ";AAAAAAARRRGH!" ), forced = "hulk")
-		dismantle_wall(1)
+		dismantle_wall(TRUE)
```
Comment (Fikou) [inline code/game/turfs/closed/walls.dm]: and here

File: code/game/turfs/closed/walls.dm [hunk 25]
```diff
@@ -269,7 +283,7 @@
 	. = ..()
 
 /turf/closed/wall/acid_melt()
-	dismantle_wall(1)
+	dismantle_wall(TRUE)
```
Comment (Fikou) [inline code/game/turfs/closed/walls.dm]: and here

**Review by Fikou — CHANGES_REQUESTED** [review]
other than that please use descript variable names in the files you created (concrete walls, concrete floors,  concrete reagent)

**Review by Fikou — APPROVED** [review]
whatever

**Pull request merged**